### PR TITLE
[db][migration] Implement db migration during container initialization

### PR DESCRIPTION
**Pull request by marcoguerri**
With support for schema migration, we'll start tracking db schema
versions and migration scripts. This implies using v0 as the base
schema version and then apply all available migrations to bring
the schema up to date. The latest version will not be tracked directly.
This patch implements this mechanism in mysql container.

### CODE REVIEW

**Comment by marcoguerri** [comment]
When the container starts up, the migration scripts are executed:

```
mysql_1    | Running database migrations...
mysql_1    | Now using version go1.13
mysql_1    | go: downloading github.com/sirupsen/logrus v1.6.0
mysql_1    | go: downloading github.com/pressly/goose v2.7.0-rc5+incompatible
mysql_1    | go: downloading github.com/go-sql-driver/mysql v1.5.0
mysql_1    | go: downloading github.com/chappjc/logrus-prefix v0.0.0-20180227015900-3a1d64819adb
mysql_1    | go: downloading golang.org/x/crypto v0.0.0-20200622213623-75b288015ac9
mysql_1    | go: downloading github.com/kballard/go-shellquote v0.0.0-20180428030007-95032a82bc51
mysql_1    | go: downloading github.com/insomniacslk/xjson v0.0.0-20190510162823-f016a4991179
mysql_1    | go: extracting github.com/sirupsen/logrus v1.6.0
mysql_1    | go: downloading golang.org/x/sys v0.0.0-20200223170610-d5e6a3e2c0ae
mysql_1    | go: extracting github.com/chappjc/logrus-prefix v0.0.0-20180227015900-3a1d64819adb
mysql_1    | go: extracting github.com/kballard/go-shellquote v0.0.0-20180428030007-95032a82bc51
mysql_1    | go: downloading github.com/mgutz/ansi v0.0.0-20170206155736-9520e82c474b
mysql_1    | go: extracting github.com/go-sql-driver/mysql v1.5.0
mysql_1    | go: extracting github.com/insomniacslk/xjson v0.0.0-20190510162823-f016a4991179
mysql_1    | go: extracting github.com/mgutz/ansi v0.0.0-20170206155736-9520e82c474b
mysql_1    | go: downloading github.com/mattn/go-colorable v0.1.6
mysql_1    | go: extracting golang.org/x/crypto v0.0.0-20200622213623-75b288015ac9
mysql_1    | go: extracting github.com/mattn/go-colorable v0.1.6
mysql_1    | go: downloading github.com/mattn/go-isatty v0.0.12
mysql_1    | go: extracting golang.org/x/sys v0.0.0-20200223170610-d5e6a3e2c0ae
mysql_1    | go: extracting github.com/mattn/go-isatty v0.0.12
mysql_1    | go: extracting github.com/pressly/goose v2.7.0-rc5+incompatible
mysql_1    | go: downloading github.com/pkg/errors v0.9.1
mysql_1    | go: extracting github.com/pkg/errors v0.9.1
mysql_1    | go: finding github.com/insomniacslk/xjson v0.0.0-20190510162823-f016a4991179
mysql_1    | go: finding github.com/chappjc/logrus-prefix v0.0.0-20180227015900-3a1d64819adb
mysql_1    | go: finding github.com/mgutz/ansi v0.0.0-20170206155736-9520e82c474b
mysql_1    | go: finding github.com/mattn/go-colorable v0.1.6
mysql_1    | go: finding github.com/mattn/go-isatty v0.0.12
mysql_1    | go: finding golang.org/x/sys v0.0.0-20200223170610-d5e6a3e2c0ae
mysql_1    | go: finding github.com/sirupsen/logrus v1.6.0
mysql_1    | go: finding golang.org/x/crypto v0.0.0-20200622213623-75b288015ac9
mysql_1    | go: finding github.com/kballard/go-shellquote v0.0.0-20180428030007-95032a82bc51
mysql_1    | go: finding github.com/go-sql-driver/mysql v1.5.0
mysql_1    | go: finding github.com/pressly/goose v2.7.0-rc5+incompatible
mysql_1    | go: finding github.com/pkg/errors v0.9.1
mysql_1    | 2020/11/29 22:42:36 OK    0001_add_extended_descriptor.sql
mysql_1    | time="2020-11-29T22:42:36Z" level=info msg="Registering target manager csvfiletargetmanager" prefix="pkg/pluginregistry"
mysql_1    | time="2020-11-29T22:42:36Z" level=info msg="Registering target manager targetlist" prefix="pkg/pluginregistry"
mysql_1    | time="2020-11-29T22:42:36Z" level=info msg="Registering test fetcher uri" prefix="pkg/pluginregistry"
mysql_1    | time="2020-11-29T22:42:36Z" level=info msg="Registering test fetcher literal" prefix="pkg/pluginregistry"
mysql_1    | time="2020-11-29T22:42:36Z" level=info msg="Registering test step echo" prefix="pkg/pluginregistry"
mysql_1    | time="2020-11-29T22:42:36Z" level=info msg="Registering test step slowecho" prefix="pkg/pluginregistry"
mysql_1    | time="2020-11-29T22:42:36Z" level=info msg="Registering test step example" prefix="pkg/pluginregistry"
mysql_1    | time="2020-11-29T22:42:36Z" level=info msg="Registering test step cmd" prefix="pkg/pluginregistry"
mysql_1    | time="2020-11-29T22:42:36Z" level=info msg="Registering test step sshcmd" prefix="pkg/pluginregistry"
mysql_1    | time="2020-11-29T22:42:36Z" level=info msg="Registering test step randecho" prefix="pkg/pluginregistry"
mysql_1    | time="2020-11-29T22:42:36Z" level=info msg="Registering reporter targetsuccess" prefix="pkg/pluginregistry"
mysql_1    | time="2020-11-29T22:42:36Z" level=info msg="Registering reporter noop" prefix="pkg/pluginregistry"
mysql_1    | 2020/11/29 22:42:36 OK    0002_migrate_descriptor_to_extended_descriptor.go
mysql_1    | 2020/11/29 22:42:36 goose: no migrations to run. current version: 2
mysql_1    | 
mysql_1    | 2020-11-29 22:42:36+00:00 [Note] [Entrypoint]: Stopping temporary server
mysql_1    | 2020-11-29T22:42:36.231864Z 12 [System] [MY-013172] [Server] Received SHUTDOWN from user root. Shutting down mysqld (Version: 8.0.21).
mysql_1    | 2020-11-29T22:42:38.261771Z 0 [System] [MY-010910] [Server] /usr/sbin/mysqld: Shutdown complete (mysqld 8.0.21)  MySQL Community Server - GPL.
mysql_1    | 2020-11-29 22:42:39+00:00 [Note] [Entrypoint]: Temporary server stopped
mysql_1    | 
mysql_1    | 2020-11-29 22:42:39+00:00 [Note] [Entrypoint]: MySQL init process done. Ready for start up.
mysql_1    | 
mysql_1    | 2020-11-29T22:42:39.524629Z 0 [System] [MY-010116] [Server] /usr/sbin/mysqld (mysqld 8.0.21) starting as process 1
mysql_1    | 2020-11-29T22:42:39.534946Z 1 [System] [MY-013576] [InnoDB] InnoDB initialization has started.
mysql_1    | 2020-11-29T22:42:40.020003Z 1 [System] [MY-013577] [InnoDB] InnoDB initialization has ended.
mysql_1    | 2020-11-29T22:42:40.178200Z 0 [System] [MY-011323] [Server] X Plugin ready for connections. Bind-address: '::' port: 33060, socket: /var/run/mysqld/mysqlx.sock
mysql_1    | 2020-11-29T22:42:40.280245Z 0 [Warning] [MY-010068] [Server] CA certificate ca.pem is self signed.
mysql_1    | 2020-11-29T22:42:40.280470Z 0 [System] [MY-013602] [Server] Channel mysql_main configured to support TLS. Encrypted connections are now supported for this channel.
mysql_1    | 2020-11-29T22:42:40.286314Z 0 [Warning] [MY-011810] [Server] Insecure configuration for --pid-file: Location '/var/run/mysqld' in the path is accessible to all OS users. Consider choosing a different directory.
mysql_1    | 2020-11-29T22:42:40.316000Z 0 [System] [MY-010931] [Server] /usr/sbin/mysqld: ready for connections. Version: '8.0.21'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  MySQL Community Server - GPL.
```

Status reports that the migrations have been executed correctly and the schema is up to date:
```
$ go run tools/migration/rdbms/main.go -dir db/rdbms/migration status
2020/11/29 22:49:27     Applied At                  Migration
2020/11/29 22:49:27     =======================================
2020/11/29 22:49:27     Sun Nov 29 22:42:36 2020 -- 0001_add_extended_descriptor.sql
2020/11/29 22:49:27     Sun Nov 29 22:42:36 2020 -- 0002_migrate_descriptor_to_extended_descriptor.go
```

**Comment by marcoguerri** [comment]
Interesting, integration tests for migration script are failing as they are being executed on v0:
```
could not migrate events in range offset 0 limit 50: could not store extended descriptor for job_id 1: Error 1054: Unknown column 'extended_descriptor' in 'field list'
```
I'll resume looking into this tomorrow, it won't affect the general approach proposed here.

**Comment by marcoguerri** [comment]
Well, the migration tool needs to run twice, against the main db and the integration tests db. Right now, it runs only against the former. Will fix tomorrow.

**Comment by tfg13** [comment]
QQ so I understand this properly: Is the idea to always create the DB with version 0 and have it migrated to latest, or is this only for containers started before, and the initial schema config will be kept up to date?

**Comment by marcoguerri** [comment]
> QQ so I understand this properly: Is the idea to always create the DB with version 0 and have it migrated to latest, or is this only for containers started before, and the initial schema config will be kept up to date?

This approach would apply for containers and any other rdbms instance supporting contest (I am writing better documentation). The idea would be to always only track the init state v0, and the schema changes to apply on top via the migration tool. There won't be a an up to date comprehensive schema config, which is also why I removed `db/rdbms/schema/latest/create_contest_db.sql` from the repo.

Integration tests will guarantee that all the known migration are applied all time to the schema and ConTest code runs against the final state of the db.

**Comment by tfg13** [comment]
Ok, was just asking, that seems fine for now the migrations on empty databases are mostly noops. If it becomes a problem we can probably still do something to have newer versions come up with the latest schema right away.

File: docker/mysql/Dockerfile [hunk 1]
```diff
@@ -1,5 +1,37 @@
 FROM mysql
```
Comment (xaionaro) [inline docker/mysql/Dockerfile:1]: Just a thought:

May be we should pick a version (to make this `Dockerfile` more stable through time).

File: docker/mysql/Dockerfile [hunk 2]
```diff
@@ -1,5 +1,37 @@
...
+RUN bash -c "bash < <(curl -s -S -L https://raw.githubusercontent.com/moovweb/gvm/master/binscripts/gvm-installer)"
+RUN bash -c "source /home/mysql/.gvm/scripts/gvm && gvm install go1.13"
+
+COPY --chown=mysql:mysql . /home/mysql/contest
+
+WORKDIR /home/mysql
+
+RUN bash -c "source .gvm/scripts/gvm && \
+             cd contest && \
+             gvm use 1.13 && gvm linkthis github.com/facebookincubator/contest"
```
Comment (xaionaro) [inline docker/mysql/Dockerfile:35]: Just out of curiosity:

Is there any reason to make that separate `RUN`-s? :)

File: docker/mysql/Dockerfile [hunk 3]
```diff
@@ -1,5 +1,37 @@
 FROM mysql
+
 # All scripts in docker-entrypoint-initdb.d/ are automatically
 # executed during container startup
 COPY docker/mysql/initdb.sql /docker-entrypoint-initdb.d/
-COPY db/rdbms/schema/latest/create_contest_db.sql /
+COPY db/rdbms/schema/v0/create_contest_db.sql /
```
Comment (xaionaro) [inline docker/mysql/Dockerfile:6]: Out of curiosity:

Why do we call it `v0`. Will there be any other version except for `0`? And why isn't it just the first migration (why do we create the initial tables in a dedicated flow)?

**Review by xaionaro — APPROVED** [review]
LGTM.

File: docker/contest/Dockerfile [hunk 4]
```diff
@@ -2,6 +2,7 @@ FROM golang:1.15.4-buster
 
 RUN apt-get update && apt-get install -y mariadb-client
 
+RUN go get -t -v ./...
```
[the remark this answers is not in the document]
Comment (marcoguerri) [reply docker/contest/Dockerfile:5]: Mh... I think this is a remainder of a previous attempt. Not sure how it slipped through in few lines diff, sorry about that. Will remove.

File: docker/mysql/Dockerfile (hunk 2 above)
Comment (marcoguerri) [reply docker/mysql/Dockerfile:35]: The previous run, i.e.
```
RUN bash -c "source /home/mysql/.gvm/scripts/gvm && gvm install go1.13"
````
is quite expensive and I'd rather cache it with as few other moving parts as possible.

File: docker/mysql/Dockerfile (hunk 3 above)
Comment (marcoguerri) [reply docker/mysql/Dockerfile:6]: > Will there be any other version except for 0?

Every migration determines a new version. I named v0 the baseline. 

> why do we create the initial tables in a dedicated flow?

It just simplifies the transition to the migration scheme, where everybody who is working with ConTest is already known to be working on the same baseline. I can make the actual v0 a migration itself. Is going to be more painful to allow people to start using the migrations, but if you think v0 is too confusing I'll just do it.

Comment (xaionaro) [reply docker/mysql/Dockerfile:6]: No-no. I was just wondering about the chain of thought. Thank you for the clarifications.

**Pull request merged**